lib/posix-process: Add per-process `sbrk`/`brk` implementations

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
Implement support for per-process `sbrk`/`brk`. This feature has two
variants: one that uses libukvmem and one that alternatively simply
allocates a big enough area from the global allocator.

In both cases, since we have to track per process the brk base and
its current position this makes the feature depend on there being an
associated `struct posix_process` with the thread upon initialization.

Additionally, one is able to choose the size of the backing memory area
by configuring `CONFIG_LIBPOSIX_PROCESS_BRK_PAGE_ORDER` (default 2MiB).

Another important note is that, by default, we adopt the behavior of
glibc with MORECORE_CLEARS == 2, meaning that whenever there is a bump
through (s)brk, we zero out that difference.
This increment zeroing behavior is configurable through
`LIBPOSIX_PROCESS_BRK_INC_MEMZERO`.

For the non-ukvmem variant we simply request a big enough area from the
global allocator and just move the brk cursor around based on requests,
while taking care to zero out on bumps.

For the ukvmem variant, we initialize an anonymous, demand mapped, VMA.
Unlike the previous variant, this one does not interfere with the global
allocator.

As far as maintaining `MORECORE_CLEARS == 2` behavior goes, every newly
faulted page will be guaranteed to be zeroed out by the page fault
handler due to the VMA not having the UK_VMA_MAP_UNINITIALIZED flag.
On one hand, this helps us in not having to memset big, page boundary
crossing, bumps all at once and, instead, have the page fault handler
zero them out as they are being used. On the other hand, we must still
manually take care of in-page bumps: if the bump does not cross a page
boundary, it will not fault and thus not be zeroed out, forcing us to do
it ourselves.

What is more, the ukvmem variant is able to free up pages upon brk
decrement requests, if the difference is a multiple of page size.
This behavior is configurable through
`LIBPOSIX_PROCESS_BRK_DEC_DONTNEED`.
This is POSIX-compliant as POSIX specifies that the pages upon
reincrement have undefined values but remains optional as it is not
Linux-compatible since there the pages do not get released and thus
reincremented pages would keep previous data (although some libcs can
do manual zeroing).

Lastly, add the declarations/signatures for `brk`/`sbrk`.

### WARNING: This depends on [elfloader #88](https://github.com/unikraft/app-elfloader/pull/88) as it deprecates its implementation.
## WARNING++: This removes/adds the `(s)brk` configurations and will ruin all catalog builds.

mogasergiu committed:
lib/posix-process: Add per-process `sbrk`/`brk` implementations

Implement support for per-process `sbrk`/`brk`. This feature has two
variants: one that uses libukvmem and one that alternatively simply
allocates a big enough area from the global allocator.

In both cases, since we have to track per process the brk base and
its current position this makes the feature depend on there being an
associated `struct posix_process` with the thread upon initialization.

Additionally, one is able to choose the size of the backing memory area
by configuring `CONFIG_LIBPOSIX_PROCESS_BRK_PAGE_ORDER` (default 2MiB).

Another important note is that, by default, we adopt the behavior of
glibc with MORECORE_CLEARS == 2, meaning that whenever there is a bump
through (s)brk, we zero out that difference.
This increment zeroing behavior is configurable through
`LIBPOSIX_PROCESS_BRK_INC_MEMZERO`.

For the non-ukvmem variant we simply request a big enough area from the
global allocator and just move the brk cursor around based on requests,
while taking care to zero out on bumps.

For the ukvmem variant, we initialize an anonymous, demand mapped, VMA.
Unlike the previous variant, this one does not interfere with the global
allocator.

As far as maintaining `MORECORE_CLEARS == 2` behavior goes, every newly
faulted page will be guaranteed to be zeroed out by the page fault
handler due to the VMA not having the UK_VMA_MAP_UNINITIALIZED flag.
On one hand, this helps us in not having to memset big, page boundary
crossing, bumps all at once and, instead, have the page fault handler
zero them out as they are being used. On the other hand, we must still
manually take care of in-page bumps: if the bump does not cross a page
boundary, it will not fault and thus not be zeroed out, forcing us to do
it ourselves.

What is more, the ukvmem variant is able to free up pages upon brk
decrement requests, if the difference is a multiple of page size.
This behavior is configurable through
`LIBPOSIX_PROCESS_BRK_DEC_DONTNEED`.
This is POSIX-compliant as POSIX specifies that the pages upon
reincrement have undefined values but remains optional as it is not
Linux-compatible since there the pages do not get released and thus
reincremented pages would keep previous data (although some libcs can
do manual zeroing).

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/nolibc: Add `brk`/`sbrk` declarations

Add the declarations/signatures for `brk`/`sbrk`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented:
Rebase force push

michpappas commented on lib/posix-process/Config.uk:
```diff
@@ -26,6 +26,23 @@ config LIBPOSIX_PROCESS_SIGNAL
+if LIBPOSIX_PROCESS_BRK
+
+config LIBPOSIX_PROCESS_BRK_NPAGES
+	int "Heap size for brk (number of pages)"
```
Since we allocate pages, perhaps it would make more sense to use page order to be consistent with Kconfig options of stacks.

michpappas commented on lib/posix-process/Config.uk:
```diff
@@ -26,6 +26,23 @@ config LIBPOSIX_PROCESS_SIGNAL
 
+menuconfig LIBPOSIX_PROCESS_BRK
+	bool "brk() system call and sbrk() libc function"
+	select LIBPOSIX_PROCESS_PIDS
```
These need to be updated because of the changes we made in `0.20.0`: 
* `CONFIG_LIBPOSIX_PROCESS_PIDS`: Replace with `CONFIG_LIBPOSIX_PROCESS_MULTITHREADING`. 
* `CONFIG_LIBPOSIX_PROCESS_INIT_PIDS`: Remove.

michpappas commented on lib/posix-process/Config.uk:
```diff
@@ -26,6 +26,23 @@ config LIBPOSIX_PROCESS_SIGNAL
+	select LIBPOSIX_PROCESS_PIDS
+	select LIBPOSIX_PROCESS_INIT_PIDS
+	select LIBUKALLOC
+	imply LIBUKVMEM
```
Not sure whether we should have this `imply LIBUKVMEM` here because of the problem that the only way to unselect these options is manually. Generally we should probably reserve `imply` for platform configs. In practice we have `imply` in a few more places (ukboot, uknetdev, posix-fdio) but most likely these shouldn't be there either.

michpappas commented on lib/posix-process/process.c:
```diff
@@ -272,6 +272,14 @@ int uk_posix_process_create(struct uk_alloc *a,
 	}
 #endif /* CONFIG_LIBPOSIX_PROCESS_SIGNAL */
 
+#if CONFIG_LIBPOSIX_PROCESS_BRK
```
Now that we raise the clone event for PID1 (b4eeed0a9443efea76b5974fc62aec515c2a03dd), the initialization & termination of brk can be moved into a `POSIX_PROCESS_CLONE_HANDLER()` in `brk.c`. That will also work well if we eventually move brk into its own library. I am also planning to remove signal initialization / termination from `process.c` too, so that we limit `process.c` to very core implementation of pprocess / pthread.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+	pprocess->brk_ctx.base = __NULL;
+
+	uk_pr_debug("Freed brk heap region: %p-%p\n",
+		    pprocess->brk_ctx.base,
```
UAF: `pprocess->brk_ctx.base` was set to `__NULL` right already. This debug message be moved up, before `free()`.

While we're there, some parts of this file use `NULL`, perhaps the implementation should be consistent and use one or the other.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+
+	rc = do_brk_inrange(cur_brk, addr_arg);
+	if (unlikely(rc))
+		return NULL;
```
I think `NULL` is not a legit return value for the `brk()`, at least the man only mentions the current break:

> C library/kernel differences
> 
> The return value described above for brk() is the behavior provided by the glibc wrapper function for the Linux brk() system call.  (On most  other  implementations,  the  return value  from  brk()  is the same; this return value was also specified in SUSv2.)  However, the actual Linux system call returns the new 
program break on success.  On failure, the system call returns the current break.  The glibc wrapper 
function does some work (i.e., checks whether the new break is less than addr) to provide the 0 and -1 return values described above.

If that is the case indeed, the check for the return value in the libc wrapper of brk and sbrk need to be updated too, as both currently checks `!rc`.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+	pprocess = uk_pprocess_current();
+	UK_ASSERT(pprocess);
+
+	base = (__uptr)pprocess->brk_ctx.base;
```
As `brk_ctx.pos` can be udpated, I think the implementation needs to protect against the case where > 1 threads in the process call `brk()`.

mogasergiu replied:
This is generally more of an SMP worry-point. However, it does no harm so I added it. There may be a problem in the future w.r.t. ukvmem so I added a comment/TODO about it.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+{
+	long rc;
+
+	rc = uk_syscall_r_brk((long)addr);
```
The syscall takes `void *`, so I wonder why the cast is needed.

mogasergiu replied:
The signatures created by the syscall define macros have all of their arguments as `long`.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+void *sbrk(intptr_t inc)
+{
+	struct posix_process *pprocess;
+	long req_base, rc;
```
`brk()` takes `void *`, so I wonder why `req_base` is declared as `long`.

mogasergiu replied:
See https://github.com/unikraft/unikraft/pull/1600#discussion_r2702526694

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+	UK_ASSERT(pprocess);
+
+	/* We are increasing or reducing our range */
+	prev_base = (__u8 *)pprocess->brk_ctx.base + pprocess->brk_ctx.pos;
```
`prev_base` should probably be `prev_brk`, as it carries the previous break value, not the base. Similarly `req_base` should be `req_brk`.

michpappas replied:
Nit: Although casting to `__u8 *` is valid, the implementation of the `brk()` syscall casts to `uptr_t` for arithmetic. It would be nice if there was some consistency across the file.

michpappas commented on lib/posix-process/brk.c:
```diff
@@ -0,0 +1,268 @@
+	 * See glibc's MORECORE_CLEARS. We go with MORECORE_CLEARS == 2 to
+	 * be on the safe side.
+	 */
+	if (addr >= cur_brk) {
```
Should that be `addr > cur_brk`? Perhaps this function should not be entered in the first place if `addr` == `cur_brk`.

michpappas reviewed (commented):
Hi @mogasergiu I think part of the implementation is just migrated from app-elfloader, but I took the opportunity to make a full review. Comments below.

mogasergiu commented:
Rebase force push.

mogasergiu committed:
lib/posix-process: Fix pthread exit event typo

Handlers were registered for the `POSIX_PROCESS_PT_EXIT_EVENT` event,
which is never raised. It's a typo, it should've been
`POSIX_PROCESS_PTEXIT_EVENT` instead.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-futex: Posixize `futex` thread exit handler

Since `futex` is a POSIX concept/feature, move its thread exit handler
to the POSIX thread exit event.

What is more this also fixes a UAF bug with GLibc's `set_tid_address`:
If the TID address is from the application's heap (so likely in the
`brk` area) and `brk` gets released before this handler runs (which
is the case as POSIX exit handlers run before generic Unikraft thread
handlers), you will get a crash.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (approved):
Thanks! 🚀 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>
Approved-by: Michalis Pappas <michalis@unikraft.io>
